# Machine Learning Project - Logistic Regression Models

--------------------------------------------

First import all 4 dataframes that we have created. After, try to model a predictor in all of them in order to see which of them gives the best results in accuracy.

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
bands_lin = pd.read_csv("dataframes/bands_lin.csv")
bands_nolin = pd.read_csv("dataframes/bands_nolin.csv")
freq_raw = pd.read_csv("dataframes/freq_raw.csv")
tonal_pitches = pd.read_csv("dataframes/tonal_pitches.csv")

In [4]:
bands_lin.head()

file  animal      band0      band1  \
0          audio-dataset/cats_dogs/cat_74.wav       1  -8.027532  -8.557745   
1  audio-dataset/cats_dogs/dog_barking_63.wav       0 -20.834045  -6.203216   
2          audio-dataset/cats_dogs/cat_60.wav       1   5.145155 -10.034791   
3  audio-dataset/cats_dogs/dog_barking_77.wav       0  -8.327550   2.926731   
4          audio-dataset/cats_dogs/cat_48.wav       1 -23.275580 -11.968597   

       band2      band3      band4      band5      band6      band7  ...  \
0  -7.066034 -11.007767  -2.205326  -2.614854  -9.251714  -7.110344  ...   
1 -13.831171 -19.378827 -28.767850 -27.528727 -26.376291 -39.701635  ...   
2  -5.291555 -21.877704 -26.527866 -33.613935 -36.843166 -37.880773  ...   
3  -4.832696  -8.023895 -18.158772 -20.334925 -23.755830 -29.417344  ...   
4  -4.606911  -2.391914 -11.736590 -31.441037 -37.686748 -37.171268  ...   

      band15     band16     band17     band18     band19     band20  \
0 -27.856952 -24.824903 -45.901381 -50.192669 -50.192669 -50.192669   
1 -37.618089 -40.776983 -42.116020 -42.455536 -42.455536 -42.455536   
2 -38.120739 -38.381084 -38.824154 -38.824154 -38.824154 -38.824154   
3 -34.919730 -36.878413 -40.819096 -41.520092 -41.520092 -41.520092   
4 -38.690999 -39.411448 -40.018420 -40.019665 -40.019665 -40.019665   

      band21     band22     band23     band24  
0 -50.192669 -50.192669 -50.192669 -50.192669  
1 -42.455536 -42.455536 -42.455536 -42.455536  
2 -38.824154 -38.824154 -38.824154 -38.824154  
3 -41.520092 -41.520092 -41.520092 -41.520092  
4 -40.019665 -40.019665 -40.019665 -40.019665  

[5 rows x 27 columns]

In [5]:
bands_nolin.head()

file  animal      band0      band1  \
0          audio-dataset/cats_dogs/cat_74.wav       1  -8.027250  -9.000595   
1  audio-dataset/cats_dogs/dog_barking_63.wav       0 -13.169041 -17.075520   
2          audio-dataset/cats_dogs/cat_60.wav       1  -2.240841 -11.906727   
3  audio-dataset/cats_dogs/dog_barking_77.wav       0  -2.627648  -5.379738   
4          audio-dataset/cats_dogs/cat_48.wav       1 -18.768084  -3.167357   

       band2      band3      band4      band5      band6      band7  ...  \
0  -4.287239  -5.630602  -9.271353  -9.189054 -16.015619 -21.270912  ...   
1 -27.049841 -28.609226 -38.392644 -34.843729 -41.878226 -41.685726  ...   
2 -28.406868 -36.586051 -38.324654 -37.770580 -38.425058 -38.553087  ...   
3 -19.012244 -23.389709 -30.632972 -36.459370 -36.834817 -38.309244  ...   
4 -16.407039 -37.284962 -36.277469 -38.758980 -39.321185 -35.662628  ...   

      band15     band16     band17     band18     band19     band20  \
0 -50.192669 -50.192669 -50.192669 -50.192669 -50.192669 -50.192669   
1 -42.455536 -42.455536 -42.455536 -42.455536 -42.455536 -42.455536   
2 -38.824154 -38.824154 -38.824154 -38.824154 -38.824154 -38.824154   
3 -41.520092 -41.520092 -41.520092 -41.520092 -41.520092 -41.520092   
4 -40.019665 -40.019665 -40.019665 -40.019665 -40.019665 -40.019665   

      band21     band22     band23     band24  
0 -50.192669 -50.192669 -50.192669 -50.192669  
1 -42.455536 -42.455536 -42.455536 -42.455536  
2 -38.824154 -38.824154 -38.824154 -38.824154  
3 -41.520092 -41.520092 -41.520092 -41.520092  
4 -40.019665 -40.019665 -40.019665 -40.019665  

[5 rows x 27 columns]

In [6]:
freq_raw.head()

file  animal      freq0      freq1  \
0          audio-dataset/cats_dogs/cat_74.wav       1   8.086168   6.813196   
1  audio-dataset/cats_dogs/dog_barking_63.wav       0 -14.020511 -17.609503   
2          audio-dataset/cats_dogs/cat_60.wav       1   9.887736  20.227963   
3  audio-dataset/cats_dogs/dog_barking_77.wav       0 -12.092080 -15.625075   
4          audio-dataset/cats_dogs/cat_48.wav       1 -30.140636 -24.407700   

       freq2      freq3      freq4      freq5      freq6      freq7  ...  \
0   0.651597  -6.359843 -10.394340 -11.307176 -14.386157 -16.761415  ...   
1 -29.791566 -31.501912 -32.045335 -29.771676 -20.327771 -11.130250  ...   
2  18.509231   1.091707   2.596950  -2.524327   6.518483  11.146668  ...   
3 -23.819625 -10.038227  -2.145106  -0.567936  -9.483675  -4.013952  ...   
4 -20.097540 -20.254327 -23.591153 -24.497132 -23.473559 -23.738149  ...   

     freq241    freq242    freq243    freq244    freq245    freq246  \
0 -50.192669 -50.192669 -50.192669 -50.192669 -50.192669 -50.192669   
1 -42.455536 -42.455536 -42.455536 -42.455536 -42.455536 -42.455536   
2 -38.824154 -38.824154 -38.824154 -38.824154 -38.824154 -38.824154   
3 -41.520092 -41.520092 -41.520092 -41.520092 -41.520092 -41.520092   
4 -40.019665 -40.019665 -40.019665 -40.019665 -40.019665 -40.019665   

     freq247    freq248    freq249    freq250  
0 -50.192669 -50.192669 -50.192669 -50.192669  
1 -42.455536 -42.455536 -42.455536 -42.455536  
2 -38.824154 -38.824154 -38.824154 -38.824154  
3 -41.520092 -41.520092 -41.520092 -41.520092  
4 -40.019665 -40.019665 -40.019665 -40.019665  

[5 rows x 253 columns]

In [7]:
tonal_pitches.head()

file  animal  chroma_0  chroma_1  \
0          audio-dataset/cats_dogs/cat_74.wav       1  0.253685  0.208958   
1  audio-dataset/cats_dogs/dog_barking_63.wav       0  0.124529  0.228619   
2          audio-dataset/cats_dogs/cat_60.wav       1  0.148572  0.224725   
3  audio-dataset/cats_dogs/dog_barking_77.wav       0  0.492111  0.460402   
4          audio-dataset/cats_dogs/cat_48.wav       1  0.045169  0.016570   

   chroma_2  chroma_3  chroma_4  chroma_5  chroma_6  chroma_7  chroma_8  \
0  0.235730  0.415372  0.626089  0.219512  0.110606  0.198507  0.249876   
1  0.484637  0.435963  0.314681  0.299143  0.345390  0.237107  0.070690   
2  0.396702  0.553182  0.355616  0.232814  0.216446  0.268361  0.313251   
3  0.265370  0.153217  0.164354  0.264637  0.371551  0.277600  0.105860   
4  0.036343  0.069753  0.150920  0.141748  0.198717  0.278540  0.339571   

   chroma_9  chroma_10  chroma_11  
0  0.163993   0.249633   0.532562  
1  0.034080   0.021759   0.055118  
2  0.486694   0.666682   0.287571  
3  0.138392   0.204018   0.324114  
4  0.432045   0.379332   0.195104

----

### Polynomial Regression

**Since our data values must be categorical, either dogs(0) or cats(1), training a polynomial regression model over our objective would result completely useless. We don't need continuous results, we only want categorical results, 0 or 1.**

**Therefore, we have to train a Logistic Regression model which returns us a value of 0 or 1.**

------

### Logistic Regression

In [64]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

#### Logistic Regression with `freq_raw` dataframe

In [65]:
X = freq_raw.drop(columns=['animal','file'], axis=1) 
y = freq_raw['animal']

X_train, X_test, y_train, y_test = train_test_split(X, y)

print(X_train.shape)
print(y_train.shape)

(207, 251)
(207,)


In [66]:
LogReg = LogisticRegression()
LogReg.fit(X_train, y_train)

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [67]:
prediction = LogReg.predict(X_test)
prediction

array([1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 1], dtype=int64)

In [68]:
LogReg.score(X_test, y_test)

0.8142857142857143

In [70]:
accuracies = []

for i in range(20):
    X_train, X_test, y_train, y_test = train_test_split(X, y)
    LogReg = LogisticRegression()
    LogReg.fit(X_train, y_train)
    prediction = LogReg.predict(X_test)
    accuracies.append(LogReg.score(X_test, y_test))

mean_accuracy = round(sum(accuracies)/len(accuracies), 5)
mean_accuracy

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_mode

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_mode

0.845

-----

#### Logistic Regression with `tonal_pitches` dataframe

In [75]:
X = tonal_pitches.drop(columns=['animal','file'], axis=1) 
y = tonal_pitches['animal']

X_train, X_test, y_train, y_test = train_test_split(X, y)

print(X_train.shape)
print(y_train.shape)

(207, 12)
(207,)


In [76]:
LogReg = LogisticRegression()
LogReg.fit(X_train, y_train)

LogisticRegression()

In [77]:
prediction = LogReg.predict(X_test)
prediction

array([1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 0, 1, 0], dtype=int64)

In [78]:
LogReg.score(X_test, y_test)

0.7142857142857143

In [79]:
accuracies = []
for i in range(10):
    X_train, X_test, y_train, y_test = train_test_split(X, y)
    LogReg = LogisticRegression()
    LogReg.fit(X_train, y_train)
    prediction = LogReg.predict(X_test)
    accuracies.append(LogReg.score(X_test, y_test))

mean_accuracy = sum(accuracies)/len(accuracies)
mean_accuracy

0.7114285714285715

#### Logistic Regression with `bands_lin` dataframe

In [80]:
X = bands_lin.drop(columns=['animal','file'], axis=1) 
y = bands_lin['animal']

X_train, X_test, y_train, y_test = train_test_split(X, y)

print(X_train.shape)
print(y_train.shape)

(207, 25)
(207,)


In [81]:
LogReg = LogisticRegression()
LogReg.fit(X_train, y_train)

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [82]:
prediction = LogReg.predict(X_test)
prediction

array([0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 1, 0, 0], dtype=int64)

In [83]:
LogReg.score(X_test, y_test)

0.8

In [84]:
accuracies = []
for i in range(10):
    X_train, X_test, y_train, y_test = train_test_split(X, y)
    LogReg = LogisticRegression()
    LogReg.fit(X_train, y_train)
    prediction = LogReg.predict(X_test)
    accuracies.append(LogReg.score(X_test, y_test))

mean_accuracy = sum(accuracies)/len(accuracies)
mean_accuracy

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_mode

0.8299999999999998

----

#### Logistic Regression with `bands_nolin` dataframe

In [85]:
X = bands_nolin.drop(columns=['animal','file'], axis=1) 
y = bands_nolin['animal']

X_train, X_test, y_train, y_test = train_test_split(X, y)

print(X_train.shape)
print(y_train.shape)

(207, 25)
(207,)


In [86]:
LogReg = LogisticRegression()
LogReg.fit(X_train, y_train)

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [87]:
prediction = LogReg.predict(X_test)
prediction

array([1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 0, 0, 0], dtype=int64)

In [88]:
LogReg.score(X_test, y_test)

0.8

In [89]:
accuracies = []
for i in range(10):
    X_train, X_test, y_train, y_test = train_test_split(X, y)
    LogReg = LogisticRegression()
    LogReg.fit(X_train, y_train)
    prediction = LogReg.predict(X_test)
    accuracies.append(LogReg.score(X_test, y_test))

mean_accuracy = sum(accuracies)/len(accuracies)
mean_accuracy

C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\polme\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_mode

0.8514285714285714

----

#### Logistic Regression with `freq_raw` dataframe accuracy $= 0.85$
#### Logistic Regression with `tonal_pitches` dataframe accuracy $= 0.71$
#### Logistic Regression with `bands_lin` dataframe accuracy $= 0.83$
#### Logistic Regression with `bands_nolin` dataframe accuracy $= 0.85$

-----

All results in accuracy are quite variable depending on the number of iterations with the logistic regression model because of the random train-test split. 

With the results obtained after many iterations, we should consider not to take into account the `tonal_pitches` dataframe.

In order to look for a more solid result, we should train our model with the `bands` and the `freq_raw` dataframes we previously created. 